### PR TITLE
Camera: Smoother collision using Raycast

### LINKED ISSUE #2276: 3rd person camera shake in confined spaces
1. dig a trench, 1 node wide, 2 nodes deep, 3 nodes long and get inside
2. toggle 3rd person view
3. walk forwards to the wall then backwards again, repeat.

this results in camera flickering when it hits the wall.

### PR DESCRIPTION
**Pull request by SmallJoker**
Fixes #2276. Currently the F7 camera "collision detection" is very rough and does not look nice at all. This PR makes it really smooth.
Code taken from my previous PR #8015

## To do

This PR is Ready for Review.

## How to test

1) F7
2) Compare smoothness of the camera collision

### CODE REVIEW

**Comment by sfan5** [comment]
Standing in a 2-node space, previously you'd get this view:
screenshot_20191216_212204

It now gives this view:
screenshot_20191216_212123
or worse this one:
screenshot_20191216_212339

**Comment by paramat** [comment]
There seem to be performance problems with using raycast, as it does some intensive object stuff that is unnecessary for the camera. See https://github.com/minetest/minetest/pull/8015#issuecomment-450529888

"i'm -1 for using raycast, as it is far more complex than the current code, and this is being used in camera update which is run every frame, so can be 60+ FPS.

Unlike the previous code, continueRaycast() detects objects, so calls getSelectedActiveObjects() which loops through all loaded objects to get those on the shootline, gets their selectionboxes, tests for a collision between the shootline and the selectionboxes and even calculates the exact intersection of the shootline with the selectionboxes. Very thorough but complete overkill for this purpose.
This is done every camera update.

Although it is thorough to account for objects it isn't needed, we have no issues with simply testing for 'walkable' and the third person camera being inside objects isn't a problem. 
The current code just tests for walkable and ignores objects."

**Comment by rubenwardy** [comment]
Our raycast is super efficient and is unlikely to cause issues with a low range limit. I suggest showing profiler results to prove this to paramat

**Comment by SmallJoker** [comment]
@paramat FYI: Raycast was added because the old code was _exponentially_ slower the higher `range` values a tool had (used in `game.cpp` IIRC). Now it's likely the best you can get in terms of speed.
I will do some profiling when I found a solution to avoid the camera to be inside the head. From what I can tell: the execution time will be about the same.

**Comment by paramat** [comment]
> Our raycast is super efficient (with only nodes)

I may have misunderstood (no one responded to my comment in the original PR), is it possible to use raycast and disable the object detection? (which is not needed for the camera).
> Raycast was added because the old code was exponentially slower the higher range values a tool had (used in game.cpp IIRC). Now it's likely the best you can get in terms of speed.

I do not doubt this, but this is not in comparison to current camera code, which may be very simple and lightweight.

My concern is that the camera collision code should perhaps be specifically written and optimised for the purpose, it may be able to be much simpler than applying a generalised raycast.
However, i have not studied this in depth.
I would appreciate some profiling, thanks. if it is not significantly more intensive i have no objection.

**Comment by rubenwardy** [comment]
> I may have misunderstood (no one responded to my comment in the original PR), is it possible to use raycast and disable the object detection? (which is not needed for the camera).

Yes. And if it were not, then it would be very easy to add such a feature

https://github.com/minetest/minetest/blob/master/src/raycast.h#L40-L41

https://github.com/minetest/minetest/blob/master/doc/lua_api.txt#L4605-L4610

**Comment by paramat** [comment]
Thanks, i also just found
https://github.com/minetest/minetest/blob/b298b0339c79db7f5b3873e73ff9ea0130f05a8a/src/raycast.h#L40

File: src/client/camera.cpp [hunk 1]
```diff
@@ -403,34 +404,37 @@ void Camera::update(LocalPlayer* player, f32 frametime, f32 busytime, f32 tool_r
...
-			if (features.walkable) {
-				my_cp.X += m_camera_direction.X*-1*-BS/2;
-				my_cp.Z += m_camera_direction.Z*-1*-BS/2;
-				my_cp.Y += m_camera_direction.Y*-1*-BS/2;
-				abort = true;
+		v3f shootline_end = my_cp - m_camera_direction * 2.75f * BS;
+		v3f shootline_extended = shootline_end - m_camera_direction * BS;
+
+		RaycastState ray(core::line3d<f32>(m_camera_position, shootline_extended),
+			true, false);
```
Comment (paramat) [inline src/client/camera.cpp]: So, how about `false, false` to remove object checking (which current code does not do)? I guess this depends on profiling.

Comment (paramat) [reply src/client/camera.cpp]: In a situation where objects are moving around near the player, perhaps it is better those objects do not move the camera about?

**Comment by paramat** [comment]
> More complex things can (and usually do) perform better than naive approaches

Yes, i should have referred only to intensiveness, not complexity/simplicity.

**Comment by paramat** [comment]
I cannot yet think of any problems caused by having the camera not collide with objects, but there is a (possibly large) performance advantage, and it seems better for objects to not move the camera around (which is current behaviour).
So i suggest that object collision is disabled, we can always enable it later if serious problems arise.

As for the change in behaviour shown by sfan5, that does not seem a big problem, as the view angle can be lowered to prevent that when the player is in a 2 node high space. I would prefer this behaviour if it measn we can use a simple raycast.

**Comment by rubenwardy** [comment]
From a quick glance, the current code doesn't check and collide with entities either

**Comment by paramat** [comment]
Function:
https://github.com/minetest/minetest/blob/b298b0339c79db7f5b3873e73ff9ea0130f05a8a/src/raycast.h#L40-L41
Call in PR:
```
RaycastState ray(core::line3d<f32>(m_camera_position, shootline_extended),
			true, false);
```
?

**Comment by SmallJoker** [comment]
Before: 5 us maximal (3 us average)
After: 1365 us maximal (600 us average)

Is this the death of this PR?

**Comment by paramat** [comment]
Wow i didn't expect that much difference. I feel a little better about making a fuss now =)
Can you confirm that object collisions are enabled? They seem to be looking at the bools in the call, if so does disabling them make a difference?
Certainly, the performance drop outweighs the benefits.

**Comment by SmallJoker** [comment]
@paramat Object detection is already disabled in commit 2/3 and 3/3.

**Comment by paramat** [comment]
Ah sorry, i did not see the new commits just before your profiling.

**Comment by SmallJoker** [comment]
There's likely no future for this PR.

**Pull request closed without merging**